### ISSUE TITLE
Binary search knot span

### ISSUE THREAD

**Issue by thibauts** [thibauts/b-spline#1]
Maybe with https://www.npmjs.com/package/binary-search-bounds

**Comment by msteinbeck** [comment]
What do you exactly mean with: Binary search knot span? Would you like to perform a binary search on the knot vector?

**Comment by thibauts** [comment]
Yes basically. The idea is to replace [this](https://github.com/thibauts/b-spline/blob/master/index.js#L41-L45).

@mikolalysenko was the originator of the idea and maybe he'll want to elaborate here.

Anyway, it would be interesting to profile the code and address any found bottleneck.

Edit: just realised you're the author of tinyspline. I'll gladly take any comment / critic, as I'm not an expert on b-splines at all.

Also, if you want to contribute code, like for knot insertion, you're more than welcome !

**Comment by msteinbeck** [comment]
When I started to implement De Boor's Algorithm I was thinking of exactly the same. I came to the conclusion that:
1. After finding the knot value `u in knots` you have to scan the left and the right side of the knot vector `knots` anyway because you need to find the the last occurrence of `u` in `knots` and count its multiplicity.
2. Lets say you have a B-Spline of degree 3 with 100 control points (we both know that B-Splines usually have not that much control points :) ), then the knot vector has 104 knots. 104 cache efficient iterations are probably not slower (maybe faster) than log2(104) cache inefficient iterations (plus scanning the left and right side as described in 1.)
3. Linear iteration gives you the ability to check for non-monotonically increasing knots vectors (u_i > u_{i+1}). To be honest TinySpline does not check the validity of the knot vector :)

So it's up to you to use binary search for the knot vector. I decided to search linearly because it simplifies the code and will not be slower unless you have a really (I mean really) big B-Spline.

> Also, if you want to contribute code, like for knot insertion, you're more than welcome !

Implementing knot insertion was a lot of work :). When I find the time, I will see if I can help you.

**Comment by thibauts** [comment]
It makes sense ... I left it this way for similar, though less elaborate, reasons. I liked the readability and would have liked to have more test cases and actual execution profiles before doing it.

Regarding knot insertions, implementing evaluation took a lot of study and mind twisting to get a clear mental picture of what the De Boor algorithm did. I wasn't able to to do the same for knot insertion after trying for a good deal of hours to get blossoms right, so I gave up. I'm not a specialist of CAGD at all so it was very tough :). That's why I asked when I saw you implemented it successfuly !

I still think it would be of great value to have this in Javascript, and have NURBS surfaces too (probably much easier), but I don't feel up to the task.

Knowing how difficult it is to wrap your head around B-splines, again, congratulations for your great work ! I was baffled when I saw tinyspline. I know how much work and thinking can go into this.

If you don't feel like adding knot insertions here, maybe a port of tinyspline to Javascript would be less time consuming and do the trick. I'd gladly help with this !

**Comment by msteinbeck** [comment]
> Knowing how difficult it is to wrap your head around B-splines, again, congratulations for your great work ! I was baffled when I saw tinyspline

Thank you :)

> If you don't feel like adding knot insertions here, maybe a port of tinyspline to Javascript would be less time consuming and do the trick. I'd gladly help with this !

TinySpline was designed to easily add bindings such as for Java and Python. Swig also supports Javascript and adding a binding for that language is feasible. Unfortunately, all bindings generated with Swig needs a shared library in order to run the binding. For Javascript that implies you have to load a dll/so/dynlib which isn't useful for simple web applications using Javascript. It is only useful if you are using something like Node.js.

So if you are aiming for a library which provides NURBS support for tools like Node.js I (we) can add a binding for Javascript. But if you would like to have a library which can be used in simple web applications then (I think) porting the code from TinySpline to your project is more eligible.

**Comment by thibauts** [comment]
Yes, it would be much more portable to have a pure Javascript implementation. It would sure be a good basis to build CAGD tools in the browser !
